### PR TITLE
One-inch mocks

### PR DESCRIPTION
**Pull request by kendricktan**
**Motivation**

Added one-inch mocks and tests

**Issue(s)**

Due to feedback from https://github.com/UMAprotocol/protocol/pull/1783

### CODE REVIEW

File: liquidator/test/common.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+const Token = artifacts.require("ExpandedERC20");
+
+const { toBN } = web3.utils;
+
+const CONSTANTS = {
+  ETH_ADDRESS: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"
+};
+
+const assertBNGreaterThan = (a, b) => {
```
Comment (chrismaree) [inline liquidator/test/common.js:9]: I'm pretty sure you can do this directly with the `gt` & `gte` directly within the bignumbers? why does it need a separate util for this action?

**Comment by nicholaspai** [comment]
> Can you provide a bit of context on these mocks? I would think that if you were using mocks now you don't need to fork main net but that doesn't seem to be the case here.

Yeah this is my main hold up I really dislike having to use mainnet-fork to test. If these are mocks that are based on the deployed 1inch contracts it'd be much cleaner IMO to test locally but I'd love to @kendricktan 's thoughts on why mainnet fork testing provides an advantage here.

File: core/contracts/common/test/OneSplitMock.sol [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        uint256 amount,
+        uint256 minReturn,
+        uint256[] memory distribution,
+        uint256 flags
+    ) public override payable returns (uint256 returnAmount) {
+        uint256 amountReturn = prices[keccak256(abi.encodePacked(fromToken, destToken))] * amount;
+
+        if (amountReturn < minReturn) {
+            revert("Min Amount not reached");
+        }
```
Comment (mrice32) [inline core/contracts/common/test/OneSplitMock.sol:49]: nit: can't this just be:
```
require(amountReturn >= minReturn, "Min Amount not reached");
```

File: liquidator/OneInchExchange.js [hunk 3]
```diff
@@ -1,39 +1,43 @@
...
+   * @param {string} swapArgs.amountWei String amount to swap, in Wei.
    * @param {Object} options Web3 options to supply to send, e.g.
    *      { from: '0x0...',
             value: '1000',
             gasPrice: '... }
    */
-  async swap(args = { fromToken, toToken, amountWei }, options = {}) {
-    // Update gasEstimator state
+  async swap({ fromToken, toToken, amountWei }, options = {}) {
+    // Implicit update to get the gasPrice :|
```
Comment (mrice32) [inline liquidator/OneInchExchange.js:40]: Was this an accidental change?

File: ci/run_truffle_tests.sh [hunk 4]
```diff
@@ -54,4 +54,4 @@ $(npm bin)/truffle test $(find ../disputer/test -name '*.js') --network ci
 $(npm bin)/truffle test $(find ../monitors/test -name '*.js') --network ci
 
 # Only run tests that needs mainnet forking here
-$(npm bin)/truffle test ../liquidator/test-fork/ --network mainnet-fork
\ No newline at end of file
+$(npm bin)/truffle test $(find ../liquidator/test-fork -name '*.js') --network mainnet-fork
```
Comment (mrice32) [inline ci/run_truffle_tests.sh:57]: Just curious -- what was the issue with the previous syntax?

File: liquidator/test/common.js (hunk 1 above)
Comment (kendricktan) [reply liquidator/test/common.js:9]: Gte and gt on chai's assert (what truffle is using) (from memory) doesnt support big integers that well. And you're right that we can just do

`assert.equal(bnA.gte(bnB))` but that wouldnt provide enough information during debugging which  is why this function exists

**Comment by kendricktan** [comment]
@chrismaree @nicholaspai

Having mainnet fork tests is really important as it gives us an indication on if our swap mechanism will work on mainnet. The mocks are useful but they're only there to test the bot logic in testnets like kovan, and is not indicative if they will work or not on mainnet, where it really matters.

Can I ask why you guys dislike testing on mainnet fork?

File: ci/run_truffle_tests.sh (hunk 4 above)
Comment (kendricktan) [reply ci/run_truffle_tests.sh:57]: Just making it consistent with the other tests

**Comment by nicholaspai** [comment]
> @chrismaree @nicholaspai
> 
> Having mainnet fork tests is really important as it gives us an indication on if our swap mechanism will work on mainnet. The mocks are useful but they're only there to test the bot logic in testnets like kovan, and is not indicative if they will work or not on mainnet, where it really matters.
> 
> What do you guys dislike when testing on mainnet forks?

Why wouldn't the swap mechanisms work on mainnet if they work on test nets? 

And I dislike it because there are some features such as automatic gas limit setting that are missing (you must explicitly set gas on some mainnet-fork transactions), and testing on it from experience is extremely slow. It's also not deterministic and is more of an integration test than a unit test.

**Comment by adrianmcli** [comment]
> Why wouldn't the swap mechanisms work on mainnet if they work on test nets?

Because they are completely different networks with different state, accounts, and in some cases consensus algorithms. With Mainnet tests, you can make sure your stuff works with real world state and liquidity. But to your point, it's very much an integration test and not a unit test. Ideally you want both in a robust system.

File: ci/run_truffle_tests.sh (hunk 4 above)
Comment (mrice32) [reply ci/run_truffle_tests.sh:57]: Gotcha. +1

**Comment by mrice32** [comment]
@kendricktan @nicholaspai @adrianmcli Great discussion! Just wanted to jot down some thoughts on this.

tldr: I think integration (mainnet-fork tests using real contracts) _and_ unit (Ganache tests using mocks) tests are valuable :).

We tend to rely more heavily on the latter since much of our protocol code does not interact with other protocols. However, when interacting with 1inch, I think an integration test, at least to validate our internal abi and interpretation of their protocol, makes sense.

> mocks are useful but they're only there to test the bot logic in testnets like kovan

I slightly disagree. I think they're _mostly_ useful for writing unit tests -- deploying them on testnets is a bonus. The idea is that they can allow you to easily set up very specific conditions in the mocked contract to test special cases and circumstances in your code. If you're testing against a mainnet contract, it can be difficult (or impossible) to replicate the conditions you want to test.

> What do you guys dislike when testing on mainnet forks?

I second what @nicholaspai said on this -- we've seen speed issues, gas estimation issues, and even behavioral differences between mainnet forks and mainnet/Ganache. It also makes tests non-hermetic and reliant on an external state and services (infura) to work.

> Ideally you want both in a robust system.

Totally agree.

I think the right call is to have both types of tests when interacting with external protocols. However, because of the many issues we've seen on mainnet forks in the past, I think we should _only_ use them where absolutely necessary (like where @kendricktan is using them for the 1inch integration) to avoid reliability and technical issues where possible.

File: core/contracts/common/test/OneSplitMock.sol [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        address fromToken,
+        address destToken,
+        uint256 amount,
+        uint256 minReturn,
+        uint256[] memory distribution,
+        uint256 flags
+    ) public override payable returns (uint256 returnAmount) {
+        uint256 amountReturn = prices[keccak256(abi.encodePacked(fromToken, destToken))] * amount;
+
+        require(amountReturn > minReturn, "Min Amount not reached");
```
[the remark this answers is not in the document]
Comment (kendricktan) [reply core/contracts/common/test/OneSplitMock.sol:47]: Ah yes, good spot :sweat_smile:

File: liquidator/test/common.js (hunk 1 above)
Comment (chrismaree) [reply liquidator/test/common.js:9]: ok cool but this is different from how we do this in other places in the repo though. Also you already have this function in the `test-fork` director [here](https://github.com/UMAprotocol/protocol/blob/0fe67182ecc811c8981b901856b1ca5ff6dc6f63/liquidator/test-fork/OneInchExchange.js#L12). If we want to have a specific function to do this then should it not be put in the common directory or more generalized test helpers? I'm just trying to think how this would play out once we need to re-use this logic elsewhere and then it's in this package.

File: liquidator/OneInchExchange.js [hunk 6]
```diff
@@ -1,38 +1,42 @@
...
   /**
    * @notice Creates OneInchExchange client
-   * @param {Object} args.web3 Web3 instance
-   * @param {Object} args.gasEstimator GasEstimator instance
+   * @param {Object} param - Constructor params
+   * @param {Object} param.web3 - Web3 instance
+   * @param {Object} param.gasEstimator - GasEstimator instance
+   * @param {string} param.oneSplitAddress - Address of the One Split
    * */
-  constructor(args = { web3, gasEstimator }) {
```
Comment (chrismaree) [inline liquidator/OneInchExchange.js:15]: I know this was here before this PR, but should the `OneInchExchange` not include the logger & associated logging states? Definitely can (and should) be a separate PR but just wanted to ask about this here :)
